tenant/service_user.go

```diff
@@ -247,6 +276,49 @@ func encryptPassword(password string) (string, error) {
 	return string(passHash), nil
 }
 
+var classes []func(rune) bool = []func(rune) bool{
+	unicode.IsNumber,
+	unicode.IsUpper,
+	unicode.IsLower,
+	func(r rune) bool {
+		found := false
+		for _, s := range errors.SpecialChars {
+			found = (s == r) || found
+		}
+		return found
+	},
+}
+
+// IsPasswordStrong checks if a password is strong enough.
+func IsPasswordStrong(password string, doCheck bool) error {
+	const numClassesRequired = 3
+	var eSlice []error = nil
+	l := len(password)
+	if l < errors.MinPasswordLen || l > errors.MaxPasswordLen {
+		eSlice = append(eSlice, errors.EPasswordLength)
+	}
+	if doCheck {
+		// make a password copy that is the length of the max password length
+		constLenPassword := strings.Repeat(password, 1+(errors.MaxPasswordLen/len(password)))[:errors.MaxPasswordLen]
+		n := 0
+
+		// Walk the whole string for each class, for constant time operation
+		for _, f := range classes {
+			found := false
+			for _, r := range constLenPassword {
+				found = f(r) || found
+			}
+			if found {
+				n++
+			}
+		}
+		if n < numClassesRequired {
+			eSlice = append(eSlice, errors.EPasswordChars)
+		}
+	}
+	return eBase.Join(eSlice...)
+}
```

Comment:
This is probably good enough to not leak useful timing information about the password.

Reply:
Putting it all together:
```
func IsPasswordStrong(password string, doCheck bool) error {
	const numClassesRequired = 3
	var eSlice []error = nil
	var tSlice []error = nil  // for constant time
	l := len(password)
	if l < errors.MinPasswordLen || > errors.MaxPasswordLen {
		eSlice = append(eSlice, errors.EPasswordLength)
	} else {
		tSlice = append(tSlice, errors.EPasswordLength)
	}

	if doCheck {
		// make a password copy that is the length of the max password length
		constLenPassword := strings.Repeat(password, 1+(errors.MaxPasswordLen/len(password)))[:errors.MaxPasswordLen]
		n := 0
		t := 0 // for constant time

		// Walk the whole string for each class, for constant time operation
		for _, f := range classes {
			found := false
			for _, r := range constLenPassword {
				found = f(r) || found
			}
			if found {
				n++
			} else {
				t++
			}
		}	
		if n < numClassesRequired {
			eSlice = append(eSlice, errors.EPasswordChars)
		} else {
			tSlice = append(tSlice, errors.EPasswordChars)
		}
	}

	eBase.Join(tSlice...) // for constant time
	return eBase.Join(eSlice...)
}
```